File: voice-call-service/src/services/ZadarmaClient.ts
import crypto from 'crypto';

/**
 * Cliente Zadarma API v1 para Node.js
 * Basado en la documentación oficial: https://github.com/zadarma/user-api-v1
 */
export class ZadarmaClient {
  private readonly apiKey: string;
  private readonly apiSecret: string;
  private readonly baseUrl: string;
  private readonly sandbox: boolean;

  constructor(apiKey: string, apiSecret: string, sandbox: boolean = false) {
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
    this.sandbox = sandbox;
    this.baseUrl = sandbox 
      ? 'https://api-sandbox.zadarma.com' 
      : 'https://api.zadarma.com';
  }

  /**
   * Genera la firma HMAC-SHA1 según la documentación oficial
   * @param signatureString - String para firmar
   * @returns Firma en base64
   */
  private encodeSignature(signatureString: string): string {
    return crypto
      .createHmac('sha1', this.apiSecret)
      .update(signatureString)
      .digest('base64');
  }

  /**
   * Genera el header de autorización según el algoritmo oficial
   * @param method - Método API (ejemplo: /v1/info/balance/)
   * @param params - Parámetros de la consulta
   * @returns Header de autorización
   */
  private getAuthHeader(method: string, params: Record<string, any>): string {
    // Filtrar parámetros que no son objetos
    const filteredParams = Object.fromEntries(
      Object.entries(params).filter(([, value]) => typeof value !== 'object')
    );

    // Ordenar parámetros alfabéticamente
    const sortedParams = Object.keys(filteredParams)
      .sort()
      .reduce((result, key) => {
        result[key] = filteredParams[key];
        return result;
      }, {} as Record<string, any>);

    // Construir query string según RFC 1738
    const paramsString = new URLSearchParams(sortedParams).toString();
    
    // Generar MD5 del query string
    const paramsMd5 = crypto.createHash('md5').update(paramsString).digest('hex');
    
    // Crear string de firma: método + parámetros + MD5(parámetros)
    const signatureString = method + paramsString + paramsMd5;
    
    // Generar firma HMAC-SHA1
    const signature = this.encodeSignature(signatureString);
    
    // Retornar header de autorización
    return `${this.apiKey}:${signature}`;
  }

  /**
   * Realiza una llamada a la API de Zadarma
   * @param method - Método API (sin /v1/, ejemplo: 'info/balance')
   * @param params - Parámetros de la consulta
   * @param requestType - Tipo de petición (GET, POST, PUT, DELETE)
   * @returns Respuesta de la API
   */
  async call(
    method: string, 
    params: Record<string, any> = {}, 
    requestType: 'GET' | 'POST' | 'PUT' | 'DELETE' = 'GET'
  ): Promise<any> {
    // Asegurar que el método comience y termine con /
    const apiMethod = `/v1/${method.replace(/^\/+|\/+$/g, '')}/`;
    
    // Agregar formato json a los parámetros
    const apiParams = { ...params, format: 'json' };
    
    // Generar header de autorización
    const authHeader = this.getAuthHeader(apiMethod, apiParams);
    
    // Construir URL y configuración de la petición
    let url = this.baseUrl + apiMethod;
    const headers: Record<string, string> = {
      'Authorization': authHeader,
      'User-Agent': 'Biosanarcall-Voice-Service/1.0',
      'Accept': 'application/json'
    };

    let body: string | undefined;

    if (requestType === 'GET') {
      // Para GET, agregar parámetros a la URL
      if (Object.keys(apiParams).length > 0) {
        url += '?' + new URLSearchParams(apiParams).toString();
      }
    } else {
      // Para POST/PUT/DELETE, enviar parámetros en el body
      headers['Content-Type'] = 'application/x-www-form-urlencoded';
      body = new URLSearchParams(apiParams).toString();
    }

    try {
      console.log(`[ZadarmaClient] ${requestType} ${url}`);
      
      const response = await fetch(url, {
        method: requestType,
        headers,
        body
      });

      const responseText = await response.text();
      
      console.log(`[ZadarmaClient] Response: ${response.status} ${responseText.substring(0, 200)}...`);

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${responseText}`);
      }

      const result = JSON.parse(responseText);

      // Verificar si la API retornó un error
      if (result.status === 'error') {
        throw new Error(`Zadarma API Error: ${result.message || 'Unknown error'}`);
      }

      return result;
    } catch (error: any) {
      console.error('[ZadarmaClient] Error:', error.message);
      throw new Error(`Zadarma API call failed: ${error.message}`);
    }
  }

  /**
   * Valida la firma de un webhook de Zadarma
   * @param postData - Datos del webhook
   * @param signature - Firma recibida en el header
   * @returns true si la firma es válida
   */
  validateWebhookSignature(postData: any, signature: string): boolean {
    try {
      let signatureString = '';
      
      // Construir string de firma según el tipo de evento
      switch (postData.event) {
        case 'NOTIFY_START':
        case 'NOTIFY_INTERNAL':
          signatureString = (postData.caller_id || '') + 
                           (postData.called_did || '') + 
                           (postData.call_start || '');
          break;
          
        case 'NOTIFY_ANSWER':
          signatureString = (postData.caller_id || '') + 
                           (postData.destination || '') + 
                           (postData.call_start || '');
          break;
          
        case 'NOTIFY_END':
          signatureString = (postData.caller_id || '') + 
                           (postData.called_did || '') + 
                           (postData.call_start || '');
          break;
          
        case 'NOTIFY_OUT_START':
        case 'NOTIFY_OUT_END':
          signatureString = (postData.internal || '') + 
                           (postData.destination || '') + 
                           (postData.call_start || '');
          break;
          
        case 'NOTIFY_RECORD':
          signatureString = (postData.pbx_call_id || '') + 
                           (postData.call_id_with_rec || '');
          break;
          
        default:
          console.warn(`[ZadarmaClient] Unknown event type: ${postData.event}`);
          return false;
      }
      
      // Generar firma esperada
      const expectedSignature = this.encodeSignature(signatureString);
      
      // Comparar firmas
      return signature === expectedSignature;
    } catch (error: any) {
      console.error('[ZadarmaClient] Signature validation error:', error.message);
      return false;
    }
  }

  // ===== MÉTODOS DE API =====

  /**
   * Obtener balance de la cuenta
   */
  async getBalance() {
    return this.call('info/balance');
  }

  /**
   * Obtener información de los números SIP
   */
  async getSip() {
    return this.call('sip');
  }

  /**
   * Obtener estado de un número SIP
   */
  async getSipStatus(sipNumber: string) {
    return this.call('sip/status', { sip: sipNumber });
  }

  /**
   * Configurar webhook para eventos
   */
  async setWebhook(webhookUrl: string) {
    return this.call('webhooks', { url: webhookUrl }, 'POST');
  }

  /**
   * Obtener estadísticas de llamadas
   */
  async getStatistics(params: {
    start?: string;
    end?: string;
    sip?: string;
    skip?: number;
    limit?: number;
  } = {}) {
    return this.call('statistics', params);
  }

  /**
   * Solicitar grabación de llamada
   */
  async getPbxRecord(callId?: string, pbxCallId?: string, lifetime?: number) {
    const params: Record<string, any> = {};
    
    if (callId) params.call_id = callId;
    if (pbxCallId) params.pbx_call_id = pbxCallId;
    if (lifetime) params.lifetime = lifetime;
    
    if (!callId && !pbxCallId) {
      throw new Error('Se requiere callId o pbxCallId');
    }
    
    return this.call('pbx/record/request', params);
  }

  /**
   * Obtener información de números directos
   */
  async getDirectNumbers() {
    return this.call('info/number');
  }

  /**
   * Configurar redirección SIP
   */
  async setSipRedirection(sipNumber: string, destination: string, enabled: boolean = true) {
    return this.call('sip/redirection', {
      sip: sipNumber,
      destination,
      status: enabled ? 'on' : 'off'
    }, 'PUT');
  }
}

File: voice-call-service/scripts/zadarma_client.py
#!/usr/bin/env python3
"""
Cliente Zadarma API v1 - Implementación exacta del algoritmo PHP oficial
Basado en: https://github.com/zadarma/user-api-v1/blob/main/lib/Client.php
"""

import requests
import hashlib
import hmac
import base64
from urllib.parse import urlencode, quote_plus
import json

class ZadarmaClient:
    """Cliente para la API de Zadarma v1"""
    
    PROD_URL = 'https://api.zadarma.com'
    SANDBOX_URL = 'https://api-sandbox.zadarma.com'
    
    def __init__(self, key=None, secret=None, is_sandbox=False):
        self.url = self.SANDBOX_URL if is_sandbox else self.PROD_URL
        # CORRECCIÓN: Secret correcto con 0 extra
        self.key = key or os.getenv('ZADARMA_KEY', '2eeea07f46fcf59e3a10')
        self.secret = secret or os.getenv('ZADARMA_SECRET', 'c87065c063195ad4b3da')
        self.http_code = None
        self.limits = {}
        
        if not self.key or not self.secret:
            raise ValueError("Zadarma API key and secret are required")
    
    def encode_signature(self, signature_string):
        """
        Generar firma HMAC-SHA1 exactamente como en la implementación que funciona
        hmac.hexdigest() -> base64.encode() (no hmac.digest() directo)
        """
        hmac_obj = hmac.new(
            self.secret.encode('utf-8'),
            signature_string.encode('utf-8'),
            hashlib.sha1
        )
        # CLAVE: Usar hexdigest() primero, luego base64
        hex_digest = hmac_obj.hexdigest().encode('utf-8')
        return base64.b64encode(hex_digest).decode('utf-8')
    
    def http_build_query(self, params):
        """
        Replicar PHP http_build_query con RFC1738 (espacios como +)
        return http_build_query($params, '', '&', PHP_QUERY_RFC1738);
        """
        if not params:
            return ''
        
        # Filtrar objetos (como en PHP)
        filtered_params = {k: v for k, v in params.items() if not hasattr(v, '__dict__')}
        
        # Ordenar por clave (como en PHP ksort)
        sorted_params = dict(sorted(filtered_params.items()))
        
        # Construir query string con RFC1738 (espacios como +)
        return urlencode(sorted_params, quote_via=quote_plus)
    
    def get_auth_header(self, method, params):
        """
        Generar header de autorización exactamente como en PHP
        """
        # Filtrar objetos y ordenar parámetros
        params_string = self.http_build_query(params)
        
        # Algoritmo exacto del PHP:
        # $signature = $this->encodeSignature($method . $paramsString . md5($paramsString));
        md5_params = hashlib.md5(params_string.encode('utf-8')).hexdigest()
        signature_string = method + params_string + md5_params
        signature = self.encode_signature(signature_string)
        
        return f"{self.key}:{signature}"
    
    def call(self, method, params=None, request_type='get', format='json'):
        """
        Hacer llamada a la API
        """
        if params is None:
            params = {}
        
        # Agregar formato
        params['format'] = format
        
        # Construir URL y headers
        url = f"{self.url}{method}"
        auth_header = self.get_auth_header(method, params)
        
        headers = {
            'Authorization': auth_header,
            'Content-Type': 'application/x-www-form-urlencoded',
            'User-Agent': 'Biosanarcall Zadarma Client/1.0'
        }
        
        try:
            request_type = request_type.lower()
            
            if request_type == 'get':
                params_string = self.http_build_query(params)
                if params_string:
                    url += f"?{params_string}"
                response = requests.get(url, headers=headers, timeout=30)
            elif request_type == 'post':
                data = self.http_build_query(params)
                response = requests.post(url, data=data, headers=headers, timeout=30)
            else:
                raise Exception(f"Método HTTP no soportado: {request_type}")
            
            self.http_code = response.status_code
            
            # Parsear rate limits de headers
            for header, value in response.headers.items():
                if header.startswith('X-RateLimit-'):
                    limit_type = header.replace('X-RateLimit-', '').lower()
                    self.limits[limit_type] = int(value)
            
            return response.text
            
        except Exception as e:
            raise Exception(f"Error en llamada API: {str(e)}")
    
    def get_balance(self):
        """Obtener balance de la cuenta"""
        return self.call('/v1/info/balance/')
    
    def get_timezone(self):
        """Obtener zona horaria"""
        return self.call('/v1/info/timezone/')
    
    def get_tariff(self):
        """Obtener información de tarifa"""
        return self.call('/v1/info/tariff/')


def test_zadarma_api():
    """Probar la API de Zadarma con el cliente"""
    
    # Credenciales CORREGIDAS
    ZADARMA_KEY = "2eeea07f46fcf59e3a10"
    ZADARMA_SECRET = "c87065c063195ad4b3da"  # Con 0 extra - CORRECCIÓN APLICADA
    
    print("🔧 Probando cliente Zadarma API v1...")
    print(f"🔑 Key: {ZADARMA_KEY}")
    
    # Crear cliente
    client = ZadarmaClient(ZADARMA_KEY, ZADARMA_SECRET)
    
    # Probar balance
    print("\n1️⃣ Probando getBalance()...")
    try:
        result = client.get_balance()
        print(f"✅ Status Code: {client.http_code}")
        print(f"📊 Response: {result}")
        
        if client.http_code == 200:
            try:
                json_result = json.loads(result)
                print(f"💰 Balance: {json.dumps(json_result, indent=2)}")
            except:
                print("⚠️ Respuesta no es JSON válido")
        else:
            print(f"❌ Error HTTP: {client.http_code}")
            
    except Exception as e:
        print(f"❌ Error: {str(e)}")
    
    # Probar timezone
    print("\n2️⃣ Probando getTimezone()...")
    try:
        result = client.get_timezone()
        print(f"✅ Status Code: {client.http_code}")
        print(f"🌍 Response: {result}")
        
    except Exception as e:
        print(f"❌ Error: {str(e)}")
    
    # Mostrar rate limits
    if client.limits:
        print(f"\n📈 Rate Limits: {client.limits}")

if __name__ == "__main__":
    test_zadarma_api()